Send / recv support in local tensor

dzmitry-huba opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #166595
    
    This change introduces LocalRunnerMode that allows you to run multiple
    SPMD functions concurrently. SMPD functions are executing one at a time,
    yielding execution capability while waiting for send or receive operations
    to complete. Send and receive peer operations only supported while running
    under LocalRunnerMode.
    
    The example test in this change demonstrates how ranks are sending data
    to the next peer and receiving data from the previous peer (ring).

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

dzmitry-huba committed:
Prototype for send / recv support in local tensor

The example here makes ranks send / recv tensors over a circle

Before send / recv
y=LocalTensor(
  0: tensor([0]),
  1: tensor([1]),
  2: tensor([2])
)

After send / recv
z=LocalTensor(
  0: tensor([2]),
  1: tensor([0]),
  2: tensor([1])
)

[ghstack-poisoned]

dzmitry-huba committed:
Update on "Prototype for send / recv support in local tensor"


The example here makes ranks send / recv tensors over a circle

Before send / recv
y=LocalTensor(
  0: tensor([0]),
  1: tensor([1]),
  2: tensor([2])
)

After send / recv
z=LocalTensor(
  0: tensor([2]),
  1: tensor([0]),
  2: tensor([1])
)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

dzmitry-huba committed:
Update on "Prototype for send / recv support in local tensor"


The example here makes ranks send / recv tensors over a circle

Before send / recv
y=LocalTensor(
  0: tensor([0]),
  1: tensor([1]),
  2: tensor([2])
)

After send / recv
z=LocalTensor(
  0: tensor([2]),
  1: tensor([0]),
  2: tensor([1])
)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

dzmitry-huba committed:
Update on "Prototype for send / recv support in local tensor"


The example here makes ranks send / recv tensors over a circle

Before send / recv
y=LocalTensor(
  0: tensor([0]),
  1: tensor([1]),
  2: tensor([2])
)

After send / recv
z=LocalTensor(
  0: tensor([2]),
  1: tensor([0]),
  2: tensor([1])
)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

dzmitry-huba committed:
Update on "Prototype for send / recv support in local tensor"


The example here makes ranks send / recv tensors over a circle

Before send / recv
y=LocalTensor(
  0: tensor([0]),
  1: tensor([1]),
  2: tensor([2])
)

After send / recv
z=LocalTensor(
  0: tensor([2]),
  1: tensor([0]),
  2: tensor([1])
)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

dzmitry-huba committed:
Update on "Prototype for send / recv support in local tensor"


The example here makes ranks send / recv tensors over a circle

Before send / recv
y=LocalTensor(
  0: tensor([0]),
  1: tensor([1]),
  2: tensor([2])
)

After send / recv
z=LocalTensor(
  0: tensor([2]),
  1: tensor([0]),
  2: tensor([1])
)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

dzmitry-huba committed:
Update on "Prototype for send / recv support in local tensor"


The example here makes ranks send / recv tensors over a circle

Before send / recv
y=LocalTensor(
  0: tensor([0]),
  1: tensor([1]),
  2: tensor([2])
)

After send / recv
z=LocalTensor(
  0: tensor([2]),
  1: tensor([0]),
  2: tensor([1])
)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

dzmitry-huba committed:
Update on "Prototype for send / recv support in local tensor"


The example here makes ranks send / recv tensors over a circle

Before send / recv
y=LocalTensor(
  0: tensor([0]),
  1: tensor([1]),
  2: tensor([2])
)

After send / recv
z=LocalTensor(
  0: tensor([2]),
  1: tensor([0]),
  2: tensor([1])
)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

dzmitry-huba committed:
Update on "Prototype for send / recv support in local tensor"


The example here makes ranks send / recv tensors over a circle

Before send / recv
y=LocalTensor(
  0: tensor([0]),
  1: tensor([1]),
  2: tensor([2])
)

After send / recv
z=LocalTensor(
  0: tensor([2]),
  1: tensor([0]),
  2: tensor([1])
)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

ezyang commented on build_variables.bzl:
```diff
@@ -1024,6 +1024,7 @@ libtorch_python_core_sources = [
 libtorch_python_distributed_core_sources = [
     "torch/csrc/distributed/c10d/init.cpp",
     "torch/csrc/distributed/c10d/python_comm_hook.cpp",
+    "torch/csrc/distributed/c10d/CallbackWork.cpp",
```
python_callback_work.cpp instead to keep in line with the other file?

ezyang commented on torch/distributed/_local_tensor/_c10d.py:
```diff
@@ -792,7 +819,63 @@ def _local_recv_any_source_(
+    return _local_recv_(tensors, process_group_so, -1, tag)
+
+
+def _unpack_ranks(ranks: int | torch.SymInt) -> list[int]:
```
I see, you want the ranks to be LocalInts. Seems like it should work.

ezyang commented on test/distributed/test_local_tensor.py:
```diff
@@ -411,5 +414,63 @@ def test_dtensor_addmm(self):
+
+            wait_all(rw)
+
+            actual[rank] = y
```
A slightly more complicated example you can do here is the 2x2 device mesh for pp and dp, run some spmd collectives on the dp submesh and then communicate across pipeline stages with pp.

dzmitry-huba replied:
Good point, rewritten the test to resemble dp, pp setup

ezyang commented on torch/distributed/_local_tensor/__init__.py:
```diff
@@ -1049,3 +1050,120 @@ def maybe_disable_local_tensor_mode() -> contextlib.AbstractContextManager:
+
+    def _run(self, id: int) -> None:
+        LocalRunnerMode.runner_context.id = id
+        # Only one thread can run at a time, hence must acquire the lock
```
i mean, technically we can run them concurrently, but for debuggability we don't (cooperative multithreading instead!)

ezyang commented on torch/csrc/distributed/c10d/CallbackWork.hpp:
```diff
@@ -0,0 +1,24 @@
+
+namespace c10d {
+
+class CallbackWork : public Work {
```
A little comment here would go a long way

dzmitry-huba replied:
Done

dzmitry-huba committed:
Update on "Send / recv support in local tensor"

    
    This change introduces LocalRunnerMode that allows you to run multiple
    SPMD functions concurrently. SMPD functions are executing one at a time,
    yielding execution capability while waiting for send or receive operations
    to complete. Send and receive peer operations only supported while running
    under LocalRunnerMode.
    
    The example test in this change demonstrates how ranks are sending data
    to the next peer and receiving data from the previous peer (ring).

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]
